### DISCUSSION TITLE
python -m build  No module named 'packaging.requirements'

### DISCUSSION THREAD

**Discussion by wb8tyw** [pypa/setuptools#3676] [Q&A] [answered]
I am stuck on this, not sure what happened from last week to this week that suddenly this is failing.
I did install some new packages, but last week I could run from a virtual environment, the "python -m build" and it was working on Anti-X / 21 (Debian bullseye based), and then stopped working this weekend.

I have rebuilt the virtual environment several times, with no change in the behavior.

Internet searches reveal that this is something that was seen around 2017 and should not be present anymore in anything close to current.

Normally the program is not run in a virtual environment, I am just using it to do the 'python -m build' as some of the python modules needed are not available as Debian packages for Anti-X.

```bash
$ python --version
Python 3.9.2
$ pip --version
pip 22.3.1 from /mnt/aviary/home/malmberg/work/d-rats/D-Rats/coyote_venv/lib/python3.9/site-packages/pip (python 3.9)

$ python -m build

Traceback (most recent call last):
  File "/mnt/aviary/home/malmberg/work/d-rats/D-Rats/coyote_venv/lib/python3.9/site-packages/build/__main__.py", line 373, in main
    built = build_call(
  File "/mnt/aviary/home/malmberg/work/d-rats/D-Rats/coyote_venv/lib/python3.9/site-packages/build/__main__.py", line 233, in build_package_via_sdist
    sdist = _build(isolation, builder, outdir, 'sdist', config_settings, skip_dependency_check)
  File "/mnt/aviary/home/malmberg/work/d-rats/D-Rats/coyote_venv/lib/python3.9/site-packages/build/__main__.py", line 143, in _build
    return _build_in_isolated_env(builder, outdir, distribution, config_settings)
  File "/mnt/aviary/home/malmberg/work/d-rats/D-Rats/coyote_venv/lib/python3.9/site-packages/build/__main__.py", line 107, in _build_in_isolated_env
    with _IsolatedEnvBuilder() as env:
  File "/mnt/aviary/home/malmberg/work/d-rats/D-Rats/coyote_venv/lib/python3.9/site-packages/build/env.py", line 99, in __enter__
    if _should_use_virtualenv():
  File "/mnt/aviary/home/malmberg/work/d-rats/D-Rats/coyote_venv/lib/python3.9/site-packages/build/env.py", line 58, in _should_use_virtualenv
    import packaging.requirements
ModuleNotFoundError: No module named 'packaging.requirements'

ERROR No module named 'packaging.requirements'
```
Selected output from pip freeze:
```text
build==0.9.0
distlib==0.3.6
packaging==21.3
pkg_resources==0.0.0
virtualenv==20.16.7
```
pyprojecdt.toml:
```toml
[build-system]
requires = ["setuptools>=40"]
build-backend = "setuptools.build_meta"

[project]
name = "D-Rats"
description = "A communications tool for D-STAR"
readme = "README.md"
requires-python = ">=3.7"
license = {text = "GPL-3"}
dynamic = ["version"]
dependencies = [
    "lxml",
    "pycairo",
    "PyGObject",
    "pyserial",
    "Pillow",
    "geopy",
    "feedparser"
]

[tool.towncrier]
directory = "changes"
package = 'd_rats.version'

[tool.setuptools]
packages = ['d_rats', 'd_rats.map', 'd_rats.sessions', 'd_rats.ui' ]

[tool.setuptools.dynamic]
version = {attr = "d_rats.version.DRATS_PEP440_VERSION"}
```

**Comment by webknjaz** [comment]
In your traceback it's visible that the exception originates in pypa/build (PEP 517 build front-end), not setuptools (PEP 517 build back-end). This is what requires the packaging library. I suppose it's missing from your system for some reason. I'd expect `build` to declare a runtime dependency on `packaging`. If it doesn't, it's their bug, if it doesn't, perhaps it's the Debian packagers' fault.
FWIW if you're exploring making use of the PEP 517 standard in debian packages, you might be interested in checking out the Fedora's pyproject RPM macros — they've got a lot of stuff around this interfere scripted already, maybe you could borrow some ideas from them.

**Comment by wb8tyw** [comment, marked as answer]
Found the issue.

I had created a packaging directory with an __init__.py in it in my project.
Once I deleted the __init__.py from the directory, the python -m build started working.

The python -m build importing that directory instead of the Pypi packaging module.

```text
$ python
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import packaging.requirements
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'packaging.requirements'
>>> import packaging
>>> dir(packaging)
['__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__']
>>> packaging.__path__
['/mnt/aviary/home/malmberg/work/d-rats/D-Rats/packaging']
>>> quit()
```

I have a python script in the packaging directory that runs towncrier and msgfmt on all the message source files since I do not know how to get the pyproject.toml to do those steps.

**Reply by webknjaz** [reply]
There's always `setup.py` and in-tree build backend possibilities.

P.S. `__init__.py` is not necessary under Python 3. It'd be the same with an empty folder too.

**Reply by wb8tyw** [reply]
It still seems that I should not be able to create a module in my project that breaks creating a package because it happens to have the same name as a module used for the build.

**Reply by webknjaz** [reply]
It's an interesting idea. Maybe you could come up with a small reproducer and file a bug? It sounds like it also may be caused by `build` but we need more info.
